### PR TITLE
Add lightning argument parser

### PR DESCRIPTION
**Pull request by nateraw**
# Before submitting

- [ ] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [ ] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [ ] Did you make sure to update the docs?   
- [ ] Did you write any new necessary tests?  

## What does this PR do?

Copy code over from #3537 to proof out the idea here before implementing directly in lightning.

## PR review    
Anyone in the community is free to review the PR once the tests have passed.     
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

**Comment by nateraw** [comment]
windows build is broken for unrelated reason it seems. we'll fix that in separate PR

File: pl_bolts/utils/arguments.py [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+    """Dataclass to represent init args of an object
+    """
+    name: str
+    types: tuple
+    default: Any
+    required: bool = False
+    context: Optional[str] = None
+
+
+class LightningArgumentParser(ArgumentParser):
```
Comment (teddykoker) [inline pl_bolts/utils/arguments.py:20]: Maybe add a docstring? Have an example of that shows the usage:
```
parser = LightningArgumentParser()
parser.add_object_args("data", MyDataModule)
parser.add_object_args("model", MyModel)
args = parser.parse_args()
```
doesn't need to be in docstring necessarily, could be in docs.

File: tests/utils/test_arguments.py [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+)
+def test_gather_lit_args(obj, expected):
+    lit_args = gather_lit_args(obj)
+    assert len(lit_args) == len(expected)
+    for lit_arg, (k, v) in zip(lit_args, expected.items()):
+        assert lit_arg.name == k
+        assert lit_arg.types[0] == v[0]
+        assert lit_arg.default == v[1]
+        assert lit_arg.required == v[2]
+
```
Comment (teddykoker) [inline tests/utils/test_arguments.py:66]: Add a test for user giving an arg that doesnt exist?

**Review by teddykoker — CHANGES_REQUESTED** [review]

File: tests/utils/test_arguments.py (hunk 2 above)
Comment (nateraw) [reply tests/utils/test_arguments.py:66]: sure

Comment (nateraw) [reply tests/utils/test_arguments.py:74]: Added simple one

**Review by teddykoker — APPROVED** [review]
Nice work Nate!

**Review by ananyahjha93 — APPROVED** [review]

**Pull request merged**